# Finding the most attractive App - Analyzing App Store and Google Play Markets

Our goal is to identify the profile of the app that is more profitable, analyzing App Store and Google Play Markets. 
This help our developers to make data-driven decisions with respect to the kind of apps are likely to attract more users. 

## Opening and Exploring Data

Considering the huge quantity of apps, we try to analyze a sample of data. In particular, we decide to analyze two existing data sets at no cost:
* a data set cointaing data about approximately ten thousand Android apps from Google Play. You can download the data set directly from [here](files/googleplaystore.csv).
* a data set cointaing data about approximately seven thousand iOs apps from App Store. You can download the data set directly from [here](files/AppleStore.csv).

In [58]:
from csv import reader 
### The Google Play data set ###
opened_file = open('C:/Users/Acer/Dataquest_projects/App Profile/googleplaystore.csv', encoding='utf8')
read_file=reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]

In [59]:
### The App Store data set ###
opened_file = open('C:/Users/Acer/Dataquest_projects/App Profile/AppleStore.csv', encoding='utf8')
read_file=reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]

We want to explore the two data sets, so we'll create a function named explore_data() that we can use repeatedly to explore rows in a more readable way. We'll also add an option for our function to show the number of rows and columns for any data set.

In [60]:
def explore_data(dataset, start, end, row_and_columns=False):
    dataset_slice=dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line between rows
        
    if row_and_columns:
        print ('Number of rows:', len(dataset))
        print ('Number of columns:', len(dataset[0]))

        
print(android_header)
print('\n')
explore_data(android,0,3,True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


The **columns** that might be uselful for our analysis are: **'App'**, **'Category'**, **'Reviews'**, **'Installs'**, **'Type'**, **'Price'** and **'Genres'**.

In [61]:
print(ios_header)
print('\n')
explore_data(ios,0,3,True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


Looking at the **columns**, we can find interesting: **"track_name"**, **"currency"**, **"price"**, **"rating_count_tot"** and **"prime_genre"**. 

We can find the details about the other columns [here](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home). 

## Deleting Wrong Data

The Google Play data set has a dedicated [discussion section](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/discussion), and we can see that [one of the discussion](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) outlines an error for row 10472. Let's print this row and compare it against the header and another row that is correct.

In [62]:
print (android[10472]) # incorrect row
print('\n')
print(android_header) # header
print('\n')
print(android[1]) # correct row

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


The row 10472 corresponds to the app *Life Made WI-Fi Touchscreen Photo Frame*, and we can see that the ratins is 19. This is clearly off because the maximum rating for a Google Play app is 5. As a consequence, we'll delete this row. 

In [63]:
print(len(android))
del android[10472] #don't run this more than once
print (len(android))

10841
10840


## Removing Duplicate Entries

If we explore the Google Play data set long enough, we'll find that some apps have more than one entry. For istance, the application Instagram has four entries.

In [64]:
for app in android:
    name=app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


In [65]:
how_many_times_app={}
for app in android:
    name=app[0]
    if name in how_many_times_app:
        how_many_times_app[name]+=1 # counts how many time the app appears
    else:
        how_many_times_app[name]=1 

# in name_app we can see haw many time every app appears 


In total, there are 1.181 cases where an app occurs more than once:

In [66]:
more_than_once=[]

times_more_than_one = 0

for app in how_many_times_app:
    if how_many_times_app[app]>1:
        more_than_once.append(app)
        times_more_than_one += how_many_times_app[app]-1 #-1 because a app has to occur only one time
            
print(times_more_than_one) # how many times apps appear more than one 

1181


We don't want to remove rows randomly, but we'll keep the rows that have the highest number of reviews because the higher the number of reviews, the more realiable the ratings.

In [67]:
reviews_max = {}

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
        
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

In a previous code cell, we found that are 1.181 cases where an app occurs more than once, so the lenght of our dictionary (of unique apps) should be equal to the difference between the lenght of our data set and 1181 

In [68]:
print ('Expected lenght',len(android)-1181)
print ('Actual lenght',len(reviews_max))

Expected lenght 9659
Actual lenght 9659



Now, let's use the reviews_max dictionary to remove the duplicates. For the duplicate cases, we'll only keep the entries with the highest number of reviews. In the code cell below:

* We start by initializing two empty lists, android_clean and already_added.
* We loop through the android data set, and for every iteration:
   * We isolate the name of the app and the number of reviews.
   * We add the current row (app) to the android_clean list, and the app name (name) to the already_added list if:
     * The number of reviews of the current app matches the number of reviews of that app as described in the reviews_max dictionary; and
     * The name of the app is not already in the already_added list. We need to add this supplementary condition to account for those cases where the highest number of reviews of a duplicate app is the same for more than one entry (for example, the Box app has three entries, and the number of reviews is the same). If we just check for reviews_max[name] == n_reviews, we'll still end up with duplicate entries for some apps.

In [69]:
android_clean = []
already_added = []

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name) # make sure this is inside the if block

Now let's quickly explore the new data set, and confirm that the number of rows is 9,659.

In [70]:
explore_data(android_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


We have 9659 rows, just as expected.

## Removing Non-English Apps

If we explore the data sets enough, we can notice that there are names of some apps that are not for an English-speaking audience.
Below, we see a couple of examples from both data sets:

In [71]:
print(ios[813][1])
print(ios[6731][1])

print(android_clean[4412][0])
print(android_clean[7940][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜
中国語 AQリスニング
لعبة تقدر تربح DZ


So, we'll remove all the apps that contains symbol that is not commonly used in English text. All the character that are specific to English texts (English alphabet, numbers, punctuation marks and symbols like +,\*,..) are encoded using the ASCII standard. 
Each ASCII character has a corresponding number between 0 and 127 associated with it, and we can take advantage of that to build a function that checks an app name and tells us whether it contains non-ASCII characters. 

In [72]:
def is_english(string):
    
    for character in string:
        if ord(character)>127:
            return False
        
    return True

In [73]:
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Instagram'))

False
True


The function seems to work fine, but some English app names use emojis or other symbols (™, — (em dash), – (en dash), etc.) that fall outside of the ASCII range. Because of this, we'll remove useful apps if we use the function in its current form.

In [74]:
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

print(ord('™'))
print(ord('😜'))

False
False
8482
128540


To minimize the impact of data loss, we'll only remove an app if its name has more than three non-ASCII characters:

In [75]:
def is_english(string):
    no_ascii = 0
    for character in string:
        if ord(character)>127:
            no_ascii += 1
            
    if no_ascii > 3:
        return False
    else:
        return True

print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


The function is still not perfect, and very few non-English apps might get past our filter, but this seems good enough at this point in our analysis. 

In [76]:
android_english =[]
ios_english = []

for app in android_clean:
    name = app[0]
    if is_english(name):
        android_english.append(app)
        
for app in ios:
    name = app[1]
    if is_english(name):
        ios_english.append(app)
        
        
explore_data(android_english,0,3,True)
print('\n')
explore_data(ios_english,0,3,True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 

We can see that we're left with 9614 Android apps and 6183 iOS apps.

## Isolating the Free Apps

As we mentioned in the introduction, we only build apps that are free to download and install. So, we need to isolate only the free apps for our analysis. 

In [77]:
android_free=[]
ios_free=[]

for app in android_english:
    price = app[7]
    if '$' in price:
        price = price.replace('$','')
    
    price = float(price)
    if price == 0:
        android_free.append(app)
        
for app in ios_english:
    price = app[4]
    if '$' in price:
        price = price.replace('$','')
    
    price = float(price)
    
    if price == 0.0 or price == 0:
        ios_free.append(app)
        
        
print(len(android_free))
print(len(ios_free))
        

8864
3222


We're left with 8864 Android apps and 3222 iOS apps, which should be enough for our analysis.

## Most Common by Genre

As we mentioned in the introduction, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

1. One Build a minimal Android version of the app, and add it to Google Play.
2. Two If the app has a good response from users, we develop it further.
3. Three If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful on both markets. For instance, a profile that works well for both markets might be a productivity app that makes use of gamification.

Let's begin the analysis by getting a sense of what are the most common genres for each market. For this, we'll need to build frequency tables for a few columns in our data sets.

We'll build two functions we can use to analyze the frequency tables:

* One function to generate frequency tables that show percentages
* Another function that we can use to display the percentages in a descending order

In [78]:
genres_android = {} # Google Play Store - dictionary where the keys are the genres and the values are the absolute frequencies
genres_ios = {}     # Apple Store - dictionary where the keys are the genres and the values are the absolute frequencies

for app in android_free:
    genre = app[9]
    if genre in genres_android:
        genres_android[genre] += 1
    else:
        genres_android[genre] = 1
        
for app in ios_free:
    genre = app[11]
    if genre in genres_ios:
        genres_ios[genre] += 1
    else:
        genres_ios[genre] = 1

table_frequency_android = {} # Google Play Store - table  of percentage frequencies for genres
table_frequency_ios = {}     # Apple Store - table  of percentage frequencies for genres

for genre in genres_android:
    table_frequency_android[genre] = 100*genres_android[genre] / len(android_free)
    
for genre in genres_ios:
    table_frequency_ios[genre] = 100*genres_ios[genre] / len(ios_free)

# tables of percentage frequencies ordered by values

table_frequency_android_ordered={k: v for k, v in sorted(table_frequency_android.items(), key=lambda item: item[1], reverse=True)}

table_frequency_ios_ordered={k: v for k, v in sorted(table_frequency_ios.items(), key=lambda item: item[1], reverse=True)}

**APPLE STORE - Table of frequency for Genre**


In [79]:
for genre in table_frequency_ios_ordered:
    print(genre + ': '+ str(table_frequency_ios_ordered[genre]) )

Games: 58.16263190564867
Entertainment: 7.883302296710118
Photo & Video: 4.9658597144630665
Education: 3.6623215394165114
Social Networking: 3.2898820608317814
Shopping: 2.60707635009311
Utilities: 2.5139664804469275
Sports: 2.1415270018621975
Music: 2.0484171322160147
Health & Fitness: 2.017380509000621
Productivity: 1.7380509000620732
Lifestyle: 1.5828677839851024
News: 1.3345747982619491
Travel: 1.2414649286157666
Finance: 1.1173184357541899
Weather: 0.8690254500310366
Food & Drink: 0.8069522036002483
Reference: 0.5586592178770949
Business: 0.5276225946617008
Book: 0.4345127250155183
Navigation: 0.186219739292365
Medical: 0.186219739292365
Catalogs: 0.12414649286157665


We can see that among the free English apps, more than a half (58.16%) are games. Entertainment apps are close to 8%, followed by photo and video apps, which are close to 5%. Only 3.66% of the apps are designed for education, followed by social networking apps which amount for 3.29% of the apps in our data set.

The general impression is that App Store (at least the part containing free English apps) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. However, the fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.

Let's continue by examining the Genres and Category columns of the Google Play data set (two columns which seem to be related).

In [80]:
category_android = {} # Google Play Store - dictionary where the keys are the categories and the values are the absolute frequencies

for app in android_free:
    category = app[1]
    if category in category_android:
        category_android[category] += 1
    else:
        category_android[category] = 1   

table_frequency_category_android = {} # Google Play Store - table  of percentage frequencies for categories

for category in category_android:
    table_frequency_category_android[category] = 100*category_android[category] / len(android_free)
    

# tables of percentage frequencies ordered by values

table_frequency_category_android_ordered={k: v for k, v in sorted(table_frequency_category_android.items(), key=lambda item: item[1], reverse=True)}


**GOOGLE PLAY STORE - Table of frequency for Category**

In [81]:
for category in table_frequency_category_android_ordered:
    print(category + ': '+ str(table_frequency_category_android_ordered[category]) )

FAMILY: 18.907942238267147
GAME: 9.724729241877256
TOOLS: 8.461191335740072
BUSINESS: 4.591606498194946
LIFESTYLE: 3.9034296028880866
PRODUCTIVITY: 3.892148014440433
FINANCE: 3.700361010830325
MEDICAL: 3.5311371841155235
SPORTS: 3.395758122743682
PERSONALIZATION: 3.3167870036101084
COMMUNICATION: 3.237815884476534
HEALTH_AND_FITNESS: 3.079873646209386
PHOTOGRAPHY: 2.9444945848375452
NEWS_AND_MAGAZINES: 2.7978339350180503
SOCIAL: 2.6624548736462095
TRAVEL_AND_LOCAL: 2.33528880866426
SHOPPING: 2.2450361010830324
BOOKS_AND_REFERENCE: 2.1435018050541514
DATING: 1.861462093862816
VIDEO_PLAYERS: 1.7937725631768953
MAPS_AND_NAVIGATION: 1.3989169675090252
FOOD_AND_DRINK: 1.2409747292418774
EDUCATION: 1.1620036101083033
ENTERTAINMENT: 0.9589350180505415
LIBRARIES_AND_DEMO: 0.9363718411552346
AUTO_AND_VEHICLES: 0.9250902527075813
HOUSE_AND_HOME: 0.8235559566787004
WEATHER: 0.8009927797833934
EVENTS: 0.7107400722021661
PARENTING: 0.6543321299638989
ART_AND_DESIGN: 0.6430505415162455
COMICS: 0.620

The landscape seems significantly different on Google Play: there are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.). However, if we investigate this further, we can see that the family category (which accounts for almost 19% of the apps) means mostly games for kids.

Even so, practical apps seem to have a better representation on Google Play compared to App Store. This picture is also confirmed by the frequency table we see for the Genres column:

**GOOGLE PLAY STORE - Table of frequency for Genre**

In [82]:
for genre in table_frequency_android_ordered:
    print(genre + ': '+ str(table_frequency_android_ordered[genre]) )

Tools: 8.449909747292418
Entertainment: 6.069494584837545
Education: 5.347472924187725
Business: 4.591606498194946
Lifestyle: 3.892148014440433
Productivity: 3.892148014440433
Finance: 3.700361010830325
Medical: 3.5311371841155235
Sports: 3.463447653429603
Personalization: 3.3167870036101084
Communication: 3.237815884476534
Action: 3.1024368231046933
Health & Fitness: 3.079873646209386
Photography: 2.9444945848375452
News & Magazines: 2.7978339350180503
Social: 2.6624548736462095
Travel & Local: 2.3240072202166067
Shopping: 2.2450361010830324
Books & Reference: 2.1435018050541514
Simulation: 2.041967509025271
Dating: 1.861462093862816
Arcade: 1.8501805054151625
Video Players & Editors: 1.7712093862815885
Casual: 1.759927797833935
Maps & Navigation: 1.3989169675090252
Food & Drink: 1.2409747292418774
Puzzle: 1.128158844765343
Racing: 0.9927797833935018
Libraries & Demo: 0.9363718411552346
Role Playing: 0.9363718411552346
Auto & Vehicles: 0.9250902527075813
Strategy: 0.9138086642599278
H

The difference between the Genres and the Category columns is not crystal clear, but one thing we can notice is that the Genres column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

## Most Popular Apps by Genre on the App Store

One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the Installs column, but for the App Store data set this information is missing. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the rating_count_tot app.

Below, we calculate the average number of user ratings per app genre on the App Store:

In [83]:
print(ios_free[9])

['343200656', 'Angry Birds', '175966208', 'USD', '0.0', '824451', '107', '4.5', '3.0', '7.4.0', '4+', 'Games', '38', '0', '10', '1']


In [84]:
ios_genre_rating={}

for app in ios_free:
    genre = app[11]
    if genre in ios_genre_rating:
        ios_genre_rating[genre] = ios_genre_rating[genre] + float(app[5])
    else:
        ios_genre_rating[genre] = float(app[5])
                
table_frequency_gen_rat={}

for genre in genres_ios:
    table_frequency_gen_rat[genre] = ios_genre_rating[genre] / genres_ios[genre]

# tables of percentage frequencies ordered by values

ios_table_frequency_gen_rat_ordered={k: v for k, v in sorted(table_frequency_gen_rat.items(), key=lambda item: item[1], reverse=True)}

for genre in ios_table_frequency_gen_rat_ordered:
    print(genre + ': '+ str(ios_table_frequency_gen_rat_ordered[genre]) )

Navigation: 86090.33333333333
Reference: 74942.11111111111
Social Networking: 71548.34905660378
Music: 57326.530303030304
Weather: 52279.892857142855
Book: 39758.5
Food & Drink: 33333.92307692308
Finance: 31467.944444444445
Photo & Video: 28441.54375
Travel: 28243.8
Shopping: 26919.690476190477
Health & Fitness: 23298.015384615384
Sports: 23008.898550724636
Games: 22788.6696905016
News: 21248.023255813954
Productivity: 21028.410714285714
Utilities: 18684.456790123455
Lifestyle: 16485.764705882353
Entertainment: 14029.830708661417
Business: 7491.117647058823
Education: 7003.983050847458
Catalogs: 4004.0
Medical: 612.0


On average, navigation apps have the highest number of user reviews, but this figure is heavily influenced by Waze and Google Maps, which have close to half a million user reviews together:

In [85]:
for app in ios_free:
    if app[-5] == 'Navigation':
        print(app[1], ':', app[5]) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


The same pattern applies to social networking apps, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc. Same applies to music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number.

Our aim is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are. The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold. We could get a better picture by removing these extremely popular apps for each genre and then rework the averages, but we'll leave this level of detail for later.

Reference apps have 74,942 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating:

In [86]:
for app in ios_free:
    if app[-5] == 'Reference':
        print(app[1], ':', app[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


However, this niche seems to show some potential. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.

This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Other genres that seem popular include weather, book, food and drink, or finance. The book genre seem to overlap a bit with the app idea we described above, but the other genres don't seem too interesting to us:

* Weather apps — people generally don't spend too much time in-app, and the chances of making profit from in-app adds are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.

* Food and drink — examples here include Starbucks, Dunkin' Donuts, McDonald's, etc. So making a popular food and drink app requires actual cooking and a delivery service, which is outside the scope of our company.

* Finance apps — these apps involve banking, paying bills, money transfer, etc. Building a finance app requires domain knowledge, and we don't want to hire a finance expert just to build an app.

Now let's analyze the Google Play market a bit.

## Most Popular Apps by Genre on Google Play

For the Google Play market, we actually have data about the number of installs, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [87]:
installs_android = {} # Google Play Store - dictionary where the keys are the genres and the values are the absolute frequencies

for app in android_free:
    installs = app[5]
    if installs in installs_android:
        installs_android[installs] += 1
    else:
        installs_android[installs] = 1

table_frequency_installs_android = {} # Google Play Store - table  of percentage frequencies for installs

for installs in installs_android:
    table_frequency_installs_android[installs] = 100*installs_android[installs] / len(android_free)
    
# tables of percentage frequencies ordered by values

table_frequency_installs_android_ordered = {k: v for k, v in sorted( table_frequency_installs_android.items(), key=lambda item: item[1], reverse=True)}

for installs in table_frequency_installs_android_ordered:
    print(installs + ': '+ str(table_frequency_installs_android_ordered[installs]) )
    


1,000,000+: 15.72653429602888
100,000+: 11.552346570397113
10,000,000+: 10.548285198555957
10,000+: 10.1985559566787
1,000+: 8.393501805054152
100+: 6.915613718411552
5,000,000+: 6.825361010830325
500,000+: 5.561823104693141
50,000+: 4.772111913357401
5,000+: 4.512635379061372
10+: 3.542418772563177
500+: 3.2490974729241877
50,000,000+: 2.3014440433212995
100,000,000+: 2.1322202166064983
50+: 1.917870036101083
5+: 0.7897111913357401
1+: 0.5076714801444043
500,000,000+: 0.27075812274368233
1,000,000,000+: 0.22563176895306858
0+: 0.04512635379061372
0: 0.01128158844765343


One problem with this data is that is not precise. For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to get an idea which app genres attract the most users, and we don't need perfect precision with respect to the number of users.

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

To perform computations, however, we'll need to convert each install number to float — this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error. We'll do this directly in the loop below, where we also compute the average number of installs for each genre (category).

In [88]:
for app in android_free:
    installs = str(app[5])
    if "+" in installs:
        installs=installs.replace('+','')
        installs=installs.replace(',','')
        app[5] = installs

android_category_installs={}

for app in android_free:
    category = app[1]
    installs = float(app[5])
    if category in android_category_installs:
        android_category_installs[category] = float(android_category_installs[category]) + installs
    else:
        android_category_installs[category] = installs
                
#print(android_genre_installs)
table_frequency_category_inst={}

for category in category_android:
    table_frequency_category_inst[category] = float(android_category_installs[category]) / float(category_android[category])

# tables of percentage frequencies ordered by values

table_frequency_category_inst_ordered={k: v for k, v in sorted(table_frequency_category_inst.items(), key=lambda item: item[1], reverse=True)}

for category in table_frequency_category_inst_ordered:
    print(category + ': '+ str(table_frequency_category_inst_ordered[category]) )

COMMUNICATION: 38456119.167247385
VIDEO_PLAYERS: 24727872.452830188
SOCIAL: 23253652.127118643
PHOTOGRAPHY: 17840110.40229885
PRODUCTIVITY: 16787331.344927534
GAME: 15588015.603248259
TRAVEL_AND_LOCAL: 13984077.710144928
ENTERTAINMENT: 11640705.88235294
TOOLS: 10801391.298666667
NEWS_AND_MAGAZINES: 9549178.467741935
BOOKS_AND_REFERENCE: 8767811.894736841
SHOPPING: 7036877.311557789
PERSONALIZATION: 5201482.6122448975
WEATHER: 5074486.197183099
HEALTH_AND_FITNESS: 4188821.9853479853
MAPS_AND_NAVIGATION: 4056941.7741935486
FAMILY: 3695641.8198090694
SPORTS: 3638640.1428571427
ART_AND_DESIGN: 1986335.0877192982
FOOD_AND_DRINK: 1924897.7363636363
EDUCATION: 1833495.145631068
BUSINESS: 1712290.1474201474
LIFESTYLE: 1437816.2687861272
FINANCE: 1387692.475609756
HOUSE_AND_HOME: 1331540.5616438356
DATING: 854028.8303030303
COMICS: 817657.2727272727
AUTO_AND_VEHICLES: 647317.8170731707
LIBRARIES_AND_DEMO: 638503.734939759
PARENTING: 542603.6206896552
BEAUTY: 513151.88679245283
EVENTS: 253542.22

On average, communication apps have the most installs: 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs:

In [89]:
print(android_free[1458])

['Drawing for Kids Learning Games for Toddlers age 3', 'FAMILY', '4.6', '29436', '43M', '1000000', 'Free', '0', 'Everyone', 'Educational;Education', 'March 21, 2018', '2.0.5.0', '4.1 and up']


In [90]:
for app in android_free:
    installs=float(app[5])
    if app[1] == 'COMMUNICATION' and (installs == 1000000000
                                      or installs == 500000000
                                      or installs == 100000000) :
        print(app[0], ':', app[5])
        

WhatsApp Messenger : 1000000000
imo beta free calls and text : 100000000
Android Messages : 100000000
Google Duo - High Quality Video Calls : 500000000
Messenger – Text and Video Chat for Free : 1000000000
imo free video calls and chat : 500000000
Skype - free IM & video calls : 1000000000
Who : 100000000
GO SMS Pro - Messenger, Free Themes, Emoji : 100000000
LINE: Free Calls & Messages : 500000000
Google Chrome: Fast & Secure : 1000000000
Firefox Browser fast & private : 100000000
UC Browser - Fast Download Private & Secure : 500000000
Gmail : 1000000000
Hangouts : 1000000000
Messenger Lite: Free Calls & Messages : 100000000
Kik : 100000000
KakaoTalk: Free Calls & Text : 100000000
Opera Mini - fast web browser : 100000000
Opera Browser: Fast and Secure : 100000000
Telegram : 100000000
Truecaller: Caller ID, SMS spam blocking & Dialer : 100000000
UC Browser Mini -Tiny Fast Private & Secure : 100000000
Viber Messenger : 500000000
WeChat : 100000000
Yahoo Mail – Stay Organized : 10000000

If we removed all the communication apps that have over 100 million installs, the average would be reduced roughly ten times:

In [91]:
under_100_m = []

for app in android_free:
    installs = float(app[5])
    
    if (app[1] == 'COMMUNICATION') and (installs < 100000000) :
        under_100_m.append(installs)
        

print(sum(under_100_m)/len(under_100_m))



3603485.3884615386


We see the same pattern for the video players category, which is the runner-up with 24,727,872 installs. The market is dominated by apps like Youtube, Google Play Movies & TV, or MX Player. The pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.).

Again, the main concern is that these app genres might seem more popular than they really are. Moreover, these niches seem to be dominated by a few giants who are hard to compete against.

The game genre seems pretty popular, but previously we found out this part of the market seems a bit saturated, so we'd like to come up with a different app recommendation if possible.

The books and reference genre looks fairly popular as well, with an average number of installs of 8,767,811. It's interesting to explore this in more depth, since we found this genre has some potential to work well on the App Store, and our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play.

Let's take a look at some of the apps from this genre and their number of installs:

In [92]:
for app in android_free:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50000
Download free book with green book : 100000
Wikipedia : 10000000
Cool Reader : 10000000
Free Panda Radio Music : 100000
Book store : 1000000
FBReader: Favorite Book Reader : 10000000
English Grammar Complete Handbook : 500000
Free Books - Spirit Fanfiction and Stories : 1000000
Google Play Books : 1000000000
AlReader -any text book reader : 5000000
Offline English Dictionary : 100000
Offline: English to Tagalog Dictionary : 500000
FamilySearch Tree : 1000000
Cloud of Books : 1000000
Recipes of Prophetic Medicine for free : 500000
ReadEra – free ebook reader : 1000000
Anonymous caller detection : 10000
Ebook Reader : 5000000
Litnet - E-books : 100000
Read books online : 5000000
English to Urdu Dictionary : 500000
eBoox: book reader fb2 epub zip : 1000000
English Persian Dictionary : 500000
Flybook : 500000
All Maths Formulas : 1000000
Ancestry : 5000000
HTC Help : 10000000
English translation from Bengali : 100000
Pdf Book Download - Read Pdf Boo


The book and reference genre includes a variety of apps: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. It seems there's still a small number of extremely popular apps that skew the average:

In [98]:
for app in android_free:
    if app[1] == 'BOOKS_AND_REFERENCE' and (float(app[5]) == 1000000000
                                            or float(app[5]) == 500000000
                                            or float(app[5]) == 100000000) :
        print(app[0], ':', app[5])

Google Play Books : 1000000000
Bible : 100000000
Amazon Kindle : 100000000
Wattpad 📖 Free Books : 100000000
Audiobooks from Audible : 100000000


However, it looks like there are only a few very popular apps, so this market still shows potential. Let's try to get some app ideas based on the kind of apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads):


In [100]:
for app in android_free:
    if app[1] == 'BOOKS_AND_REFERENCE' and (float(app[5]) == 1000000
                                            or float(app[5]) == 5000000
                                            or float(app[5]) == 10000000
                                            or float(app[5]) == 50000000):
        print(app[0], ':', app[5])

Wikipedia : 10000000
Cool Reader : 10000000
Book store : 1000000
FBReader: Favorite Book Reader : 10000000
Free Books - Spirit Fanfiction and Stories : 1000000
AlReader -any text book reader : 5000000
FamilySearch Tree : 1000000
Cloud of Books : 1000000
ReadEra – free ebook reader : 1000000
Ebook Reader : 5000000
Read books online : 5000000
eBoox: book reader fb2 epub zip : 1000000
All Maths Formulas : 1000000
Ancestry : 5000000
HTC Help : 10000000
Moon+ Reader : 10000000
English-Myanmar Dictionary : 1000000
Golden Dictionary (EN-AR) : 1000000
All Language Translator Free : 1000000
Aldiko Book Reader : 10000000
Dictionary - WordWeb : 5000000
50000 Free eBooks & Free AudioBooks : 5000000
Al-Quran (Free) : 10000000
Al Quran Indonesia : 10000000
Al'Quran Bahasa Indonesia : 10000000
Al Quran Al karim : 1000000
Al Quran : EAlim - Translations & MP3 Offline : 5000000
Koran Read &MP3 30 Juz Offline : 1000000
Hafizi Quran 15 lines per page : 1000000
Quran for Android : 10000000
Satellite AR : 

This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

We also notice there are quite a few apps built around the book Quran, which suggests that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets.

However, it looks like the market is already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.

## Conclusions

In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.